File: donater-bot/utils/tg_api.py
"""Raw Telegram HTTP JSON — inline-кнопки с полем style."""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Dict, Optional

import aiohttp

from config import config

log = logging.getLogger(__name__)


def _base() -> str:
    return f"https://api.telegram.org/bot{config.bot_token}"


async def raw_send_message(
    chat_id: int,
    text: str,
    reply_markup: Optional[Dict[str, Any]] = None,
    parse_mode: str = "HTML",
) -> dict:
    payload: Dict[str, Any] = {
        "chat_id": chat_id,
        "text": text,
        "parse_mode": parse_mode,
    }
    if reply_markup is not None:
        payload["reply_markup"] = reply_markup
    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{_base()}/sendMessage",
                json=payload,
                timeout=aiohttp.ClientTimeout(total=60),
            ) as resp:
                data = await resp.json(content_type=None)
                if not data.get("ok"):
                    log.error("sendMessage: %s", data)
                return data
    except Exception as exc:
        log.exception("raw_send_message: %s", exc)
        return {"ok": False, "description": str(exc)}


async def raw_send_photo_file(
    chat_id: int,
    file_path: Path,
    caption: str,
    reply_markup: Optional[Dict[str, Any]] = None,
    parse_mode: str = "HTML",
) -> dict:
    try:
        file_bytes = file_path.read_bytes()
    except OSError as exc:
        log.error("read photo: %s", exc)
        return {"ok": False, "description": str(exc)}
    data = aiohttp.FormData()
    data.add_field("chat_id", str(chat_id))
    data.add_field("caption", caption)
    data.add_field("parse_mode", parse_mode)
    data.add_field(
        "photo",
        file_bytes,
        filename=file_path.name,
        content_type="image/jpeg",
    )
    if reply_markup is not None:
        data.add_field("reply_markup", json.dumps(reply_markup))
    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{_base()}/sendPhoto",
                data=data,
                timeout=aiohttp.ClientTimeout(total=120),
            ) as resp:
                body = await resp.json(content_type=None)
                if not body.get("ok"):
                    log.error("sendPhoto: %s", body)
                return body
    except Exception as exc:
        log.exception("raw_send_photo_file: %s", exc)
        return {"ok": False, "description": str(exc)}


async def raw_send_photo_file_id(
    chat_id: int,
    file_id: str,
    caption: str,
    reply_markup: Optional[Dict[str, Any]] = None,
    parse_mode: str = "HTML",
) -> dict:
    payload: Dict[str, Any] = {
        "chat_id": chat_id,
        "photo": file_id,
        "caption": caption,
        "parse_mode": parse_mode,
    }
    if reply_markup is not None:
        payload["reply_markup"] = reply_markup
    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{_base()}/sendPhoto",
                json=payload,
                timeout=aiohttp.ClientTimeout(total=60),
            ) as resp:
                return await resp.json(content_type=None)
    except Exception as exc:
        log.exception("raw_send_photo_file_id: %s", exc)
        return {"ok": False}


async def raw_edit_reply_markup(
    chat_id: int, message_id: int, reply_markup: Optional[Dict[str, Any]]
) -> dict:
    payload: Dict[str, Any] = {
        "chat_id": chat_id,
        "message_id": message_id,
        "reply_markup": reply_markup or {"inline_keyboard": []},
    }
    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{_base()}/editMessageReplyMarkup",
                json=payload,
                timeout=aiohttp.ClientTimeout(total=30),
            ) as resp:
                return await resp.json(content_type=None)
    except Exception as exc:
        log.exception("raw_edit_reply_markup: %s", exc)
        return {"ok": False}


def style_btn(text: str, callback_data: str, style: Optional[str] = None) -> dict:
    """Только text + callback_data. Поле ``style`` в Bot API для inline даёт 400
    (invalid button style) у части клиентов/версий API — не отправляем."""
    return {"text": text, "callback_data": callback_data}


def main_menu_markup(lang: str) -> dict:
    """Игры подставляются в хендлере после загрузки из БД — здесь шаблон без id."""
    return {
        "inline_keyboard": [
            [
                style_btn("🔥 Free Fire", "nav:game:__FF__", "destructive"),
                style_btn("🎮 PUBG Mobile", "nav:game:__PUBG__", "primary"),
            ],
            [
                {"text": "⚙️ Настройки", "callback_data": "menu:settings"},
                {"text": "📞 Поддержка", "url": config.support_url},
            ],
        ]
    }


def build_main_menu_from_games(rows_game_buttons: list) -> dict:
    """rows_game_buttons: list of rows, each row list of dict buttons."""
    kb = rows_game_buttons + [
        [
            {"text": "⚙️ Настройки", "callback_data": "menu:settings"},
            {"text": "📞 Поддержка", "url": config.support_url},
        ]
    ]
    return {"inline_keyboard": kb}


def product_keyboard_rows(products: list, back_cb: str = "menu:home") -> dict:
    rows = []
    for p in products:
        st = "primary"
        if p["is_popular"]:
            st = "destructive"
        elif p["is_best_value"]:
            st = "success"
        fire = "🔥 " if p["is_popular"] else ""
        star = "⭐ " if p["is_best_value"] else ""
        line = f"{fire}{star}{p['label']} — {p['price_tjs']} смн"[:64]
        rows.append([style_btn(line, f"prod:{p['id']}", st)])
    rows.append([{"text": "◀️ Назад", "callback_data": back_cb}])
    return {"inline_keyboard": rows}


def payment_methods_markup(
    t_dc: str, t_alif: str, t_mc: str, t_milli: str, t_back: str
) -> dict:
    return {
        "inline_keyboard": [
            [style_btn(t_dc, "pay:dc_city", "primary")],
            [style_btn(t_alif, "pay:alif", "success")],
            [style_btn(t_mc, "pay:mastercard", "destructive")],
            [style_btn(t_milli, "pay:milli", "primary")],
            [{"text": t_back, "callback_data": "flow:back_summary"}],
        ]
    }


def confirm_order_markup(t_yes: str, t_no: str) -> dict:
    return {
        "inline_keyboard": [
            [
                style_btn(t_yes, "order:confirm_yes", "success"),
                style_btn(t_no, "order:confirm_no", "destructive"),
            ]
        ]
    }


def admin_order_actions(t_accept: str, t_reject: str, t_write: str, oid: int, uid: int) -> dict:
    return {
        "inline_keyboard": [
            [
                style_btn(t_accept, f"adm:acc:{oid}", "success"),
                style_btn(t_reject, f"adm:rej:{oid}", "destructive"),
            ],
            [{"text": t_write, "url": f"tg://user?id={uid}"}],
        ]
    }


def captcha_keyboard(salt: str, opts: list) -> dict:
    rows = []
    for opt in opts:
        st = "success" if opt["correct"] else "destructive"
        rows.append(
            [style_btn(str(opt["label"]), f"cap:{salt}:{opt['value']}", st)]
        )
    return {"inline_keyboard": rows}


def review_admin_keyboard_raw(review_id: int, t) -> dict:
    return {
        "inline_keyboard": [
            [
                style_btn(t("review_btn_add"), f"revok:{review_id}", "success"),
                style_btn(t("review_btn_edit"), f"reved:{review_id}", "primary"),
            ],
            [style_btn(t("review_btn_reject"), f"revno:{review_id}", "destructive")],
        ]
    }

File: donater-bot/utils/tg_api.js
/**
 * Raw Telegram Bot API calls (JSON + multipart) so inline button `style`
 * is sent without Grammy stripping/validating unknown fields.
 */
import { createReadStream } from 'fs';
import { existsSync } from 'fs';
import { basename } from 'path';
import FormData from 'form-data';
import { config } from '../config.js';

function baseUrl() {
  return `https://api.telegram.org/bot${config.botToken}`;
}

async function parseResponse(res) {
  const text = await res.text();
  let json = {};
  try {
    json = text ? JSON.parse(text) : {};
  } catch {
    json = { ok: false, description: text };
  }
  if (!res.ok || json.ok === false) {
    console.error('tg_api error', res.status, json.description || json);
  }
  return json;
}

/**
 * @param {number|string} chatId
 * @param {string} text
 * @param {{ parse_mode?: string, reply_markup?: object, disable_web_page_preview?: boolean }} [opts]
 */
export async function sendMessageRaw(chatId, text, opts = {}) {
  const body = {
    chat_id: chatId,
    text,
    parse_mode: opts.parse_mode ?? 'HTML',
  };
  if (opts.reply_markup) body.reply_markup = opts.reply_markup;
  if (opts.disable_web_page_preview != null) {
    body.disable_web_page_preview = opts.disable_web_page_preview;
  }
  const res = await fetch(`${baseUrl()}/sendMessage`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return parseResponse(res);
}

/**
 * @param {number|string} chatId
 * @param {string} photo file_id or local filesystem path
 * @param {string} caption
 * @param {{ parse_mode?: string, reply_markup?: object }} [opts]
 */
export async function sendPhotoRaw(chatId, photo, caption, opts = {}) {
  const parseMode = opts.parse_mode ?? 'HTML';
  const replyMarkup = opts.reply_markup;

  const isLocalPath =
    typeof photo === 'string' &&
    photo.length > 0 &&
    (photo.includes('/') || photo.includes('\\') || photo.startsWith('file:'));

  if (!isLocalPath) {
    const body = {
      chat_id: chatId,
      photo,
      caption,
      parse_mode: parseMode,
    };
    if (replyMarkup) body.reply_markup = replyMarkup;
    const res = await fetch(`${baseUrl()}/sendPhoto`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    return parseResponse(res);
  }

  if (!existsSync(photo)) {
    console.error('tg_api sendPhotoRaw: file missing', photo);
    return sendMessageRaw(chatId, caption || '', { ...opts, parse_mode: parseMode });
  }

  const form = new FormData();
  form.append('chat_id', String(chatId));
  form.append('photo', createReadStream(photo), { filename: basename(photo) });
  form.append('caption', caption ?? '');
  form.append('parse_mode', parseMode);
  if (replyMarkup) {
    form.append('reply_markup', JSON.stringify(replyMarkup));
  }
  const res = await fetch(`${baseUrl()}/sendPhoto`, {
    method: 'POST',
    body: form,
    headers: form.getHeaders(),
  });
  return parseResponse(res);
}

/**
 * @param {number|string} chatId
 * @param {number} messageId
 * @param {string} text
 * @param {{ parse_mode?: string, reply_markup?: object }} [opts]
 */
export async function editMessageTextRaw(chatId, messageId, text, opts = {}) {
  const body = {
    chat_id: chatId,
    message_id: messageId,
    text,
    parse_mode: opts.parse_mode ?? 'HTML',
  };
  if (opts.reply_markup) body.reply_markup = opts.reply_markup;
  const res = await fetch(`${baseUrl()}/editMessageText`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return parseResponse(res);
}

/**
 * Admin order / review inline keyboard with styled buttons (raw JSON).
 */
export function adminOrderActionKeyboard(i18n, orderId, userTgId) {
  return {
    inline_keyboard: [
      [
        { text: i18n('btn_accept'), callback_data: `adm:acc:${orderId}`, style: 'success' },
        { text: i18n('btn_reject'), callback_data: `adm:rej:${orderId}`, style: 'destructive' },
      ],
      [{ text: i18n('btn_write_user'), url: `tg://user?id=${userTgId}` }],
    ],
  };
}

export function reviewAdminKeyboardRaw(reviewId, t) {
  return {
    inline_keyboard: [
      [
        { text: t('review_btn_add'), callback_data: `revok:${reviewId}`, style: 'success' },
        { text: t('review_btn_edit'), callback_data: `reved:${reviewId}`, style: 'primary' },
      ],
      [{ text: t('review_btn_reject'), callback_data: `revno:${reviewId}`, style: 'destructive' }],
    ],
  };
}
